src/main/java/graphql/execution/instrumentation/Instrumentation.java

```diff
@@ -72,62 +71,26 @@ default InstrumentationState createState() {
-     */
-    default InstrumentationContext<CompletableFuture<ExecutionResult>> beginDataFetchDispatch(InstrumentationDataFetchParameters parameters) {
-        return (result, t) -> {
-        };
-    }
-
-    /**
-     * This is called just before the data fetching stage is started, waits for all data to be completed and when this step finishes the {@link InstrumentationContext#onEnd(Object, Throwable)}
-     * will be called indicating that the step has finished.
+     * This is called just before the execution of the query operation is started.
      *
      * @param parameters the parameters to this step
      *
      * @return a non null {@link InstrumentationContext} object that will be called back when the step ends
      */
-    InstrumentationContext<ExecutionResult> beginDataFetch(InstrumentationDataFetchParameters parameters);
+    InstrumentationContext<ExecutionResult> beginExecuteOperation(InstrumentationExecuteOperationParameters parameters);
 
     /**
-     * This is called each time the {@link graphql.execution.ExecutionStrategy} is invoked and when the
-     * {@link java.util.concurrent.CompletableFuture} has been dispatched for the query fields the
-     * {@link graphql.execution.instrumentation.InstrumentationContext#onEnd(Object, Throwable)}
-     * is called.
-     *
-     * Note because the execution strategy execution is asynchronous, the query data is not guaranteed to be
-     * completed when this step finishes.  It is however a chance to dispatch side effects that might cause
-     * asynchronous data fetching code to actually run or attach CompletableFuture handlers onto the result
-     * via Instrumentation.
+     * This is called each time an {@link graphql.execution.ExecutionStrategy} is invoked, which may be multiple times
+     * per query as the engine recursively descends down over the query.
      *
      * @param parameters the parameters to this step
      *
      * @return a non null {@link InstrumentationContext} object that will be called back when the step ends
      */
-    InstrumentationContext<CompletableFuture<ExecutionResult>> beginExecutionStrategy(InstrumentationExecutionStrategyParameters parameters);
+    InstrumentationContext<ExecutionResult> beginExecutionStrategy(InstrumentationExecutionStrategyParameters parameters);
 
     /**
-     * This is called just before a selection set of fields is resolved and when this step finishes the {@link InstrumentationContext#onEnd(Object, Throwable)}
-     * will be called indicating that the step has finished.
-     *
-     * @param parameters the parameters to this step
-     *
-     * @return a non null {@link InstrumentationContext} object that will be called back when the step ends
-     */
-    default InstrumentationContext<Map<String, List<Field>>> beginFields(InstrumentationExecutionStrategyParameters parameters) {
-        return (result, t) -> {
-        };
-    }
-
```

Comment:
This was always called at the same time as `beginExecutionStrategy` so I merged them.